runtime/str.go

```diff
@@ -453,6 +453,65 @@ func strNew(f *Frame, t *Type, args Args, _ KWArgs) (*Object, *BaseException) {
 	return s.ToObject(), nil
 }
 
+// strReplace returns a copy of the string s with the first n non-overlapping
+// instances of old replaced by sub. If old is empty, it matches at the
+// beginning of the string. If n < 0, there is no limit on the number of
+// replacements.
+func strReplace(f *Frame, args Args, _ KWArgs) (*Object, *BaseException) {
+	var raised *BaseException
+	// TODO: Support unicode replace.
+	expectedTypes := []*Type{StrType, StrType, StrType, ObjectType}
+	argc := len(args)
+	if argc == 3 {
+		expectedTypes = expectedTypes[:argc]
+	}
+	if raised := checkMethodArgs(f, "replace", args, expectedTypes...); raised != nil {
+		return nil, raised
+	}
+	n := -1
+	if argc == 4 {
+		n, raised = ToIntValue(f, args[3])
+		if raised != nil {
+			return nil, raised
+		}
+	}
+	s := toStrUnsafe(args[0]).Value()
+	old := toStrUnsafe(args[1]).Value()
+	sub := toStrUnsafe(args[2]).Value()
+	// CPython only, pypy supposed to be same as Go.
+	if len(s) == 0 && len(old) == 0 && n >= 0 {
+		return NewStr("").ToObject(), nil
+	}
+	if n == 0 {
+		return NewStr(s).ToObject(), nil
+	}
+	if len(old) == 0 {
+		if n < 0 {
+			n = len(s) + 1
+		}
+		buf := bytes.Buffer{}
+		if n > 0 {
+			buf.WriteString(sub)
+			n--
+		}
+		for i := 0; i < len(s); i++ {
+			buf.WriteByte(s[i])
+			if n > 0 {
+				buf.WriteString(sub)
+				n--
+			} else {
+				i++
+				if i < len(s) {
```

Comment:
Can we do this branch after the loop? I think the code would read more clearly: do the required replacements, then write the remainder of the string.

I think things would be tidier if your loop condition dictated when you exit the loop:

```
i := 0
for n > 0 && i < numBytes {
  // .. replacement logic
  i++
  n--
}
```

As written it's not immediately obvious what the termination condition is.